In [1]:
import pandas as pd
import warnings
warnings.simplefilter("ignore")
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
users = pd.read_csv("takehome_users.csv",encoding = 'latin1')
users['last_session_creation_time'] = pd.to_datetime(users['last_session_creation_time'], unit='s')
users.head(10)

object_id        creation_time               name  \
0          1  2014-04-22 03:53:30     Clausen August   
1          2  2013-11-15 03:45:04      Poole Matthew   
2          3  2013-03-19 23:14:52  Bottrill Mitchell   
3          4  2013-05-21 08:09:28    Clausen Nicklas   
4          5  2013-01-17 10:14:20          Raw Grace   
5          6  2013-12-17 03:37:06      Cunha Eduardo   
6          7  2012-12-16 13:24:32       Sewell Tyler   
7          8  2013-07-31 05:34:02  Hamilton Danielle   
8          9  2013-11-05 04:04:24         Amsel Paul   
9         10  2013-01-16 22:08:03       Santos Carla   

                           email    creation_source  \
0       AugustCClausen@yahoo.com       GUEST_INVITE   
1         MatthewPoole@gustr.com         ORG_INVITE   
2     MitchellBottrill@gustr.com         ORG_INVITE   
3      NicklasSClausen@yahoo.com       GUEST_INVITE   
4             GraceRaw@yahoo.com       GUEST_INVITE   
5  EduardoPereiraCunha@yahoo.com       GUEST_INVITE   
6     TylerSewell@jourrapide.com             SIGNUP   
7     DanielleHamilton@yahoo.com  PERSONAL_PROJECTS   
8          PaulAmsel@hotmail.com  PERSONAL_PROJECTS   
9  CarlaFerreiraSantos@gustr.com         ORG_INVITE   

  last_session_creation_time  opted_in_to_mailing_list  \
0        2014-04-22 03:53:30                         1   
1        2014-03-31 03:45:04                         0   
2        2013-03-19 23:14:52                         0   
3        2013-05-22 08:09:28                         0   
4        2013-01-22 10:14:20                         0   
5        2013-12-19 03:37:06                         0   
6        2012-12-20 13:24:32                         0   
7                        NaT                         1   
8                        NaT                         0   
9        2014-06-03 22:08:03                         1   

   enabled_for_marketing_drip  org_id  invited_by_user_id  
0                           0      11             10803.0  
1                           0       1               316.0  
2                           0      94              1525.0  
3                           0       1              5151.0  
4                           0     193              5240.0  
5                           0     197             11241.0  
6                           1      37                 NaN  
7                           1      74                 NaN  
8                           0     302                 NaN  
9                           1     318              4143.0

In [3]:
engage = pd.read_csv("takehome_user_engagement.csv")
engage = pd.read_csv('takehome_user_engagement.csv', encoding='latin-1', parse_dates=['time_stamp'])
engage.head()

time_stamp  user_id  visited
0 2014-04-22 03:53:30        1        1
1 2013-11-15 03:45:04        2        1
2 2013-11-29 03:45:04        2        1
3 2013-12-09 03:45:04        2        1
4 2013-12-25 03:45:04        2        1

In [4]:
engage.shape

(207917, 3)

In [5]:
#Sort "engage" by users' number of visits
import time
#Num_visits = engage["user_id"].value_counts()
#adopted = Num_visits[Num_visits >= 3]
#weeks
#adopted.count()


In [6]:
import time
engage['week'] = engage['time_stamp'].dt.strftime('%Y-%U')

weeks = engage.groupby(['user_id', 'week']).size().reset_index().rename(columns={0:'count'})
weeks.head()

user_id     week  count
0        1  2014-16      1
1        2  2013-45      1
2        2  2013-47      1
3        2  2013-49      1
4        2  2013-51      1

In [7]:
adopted = weeks[weeks['count'] >= 3].user_id.unique()
adopted1 = weeks[weeks['count'] >= 3]
#adopted.head()

In [8]:
adopted1

user_id     week  count
23          10  2013-08      3
26          10  2013-11      3
28          10  2013-14      3
29          10  2013-15      3
30          10  2013-16      3
31          10  2013-17      4
32          10  2013-18      4
33          10  2013-19      4
36          10  2013-22      3
37          10  2013-23      5
38          10  2013-24      4
40          10  2013-26      5
41          10  2013-27      3
42          10  2013-28      4
43          10  2013-29      4
44          10  2013-30      5
45          10  2013-31      4
47          10  2013-33      4
48          10  2013-34      5
49          10  2013-35      4
50          10  2013-36      5
51          10  2013-37      3
52          10  2013-38      3
53          10  2013-39      3
54          10  2013-40      3
55          10  2013-41      3
56          10  2013-42      4
57          10  2013-43      3
59          10  2013-45      3
62          10  2013-48      3
...        ...      ...    ...
69579    11975  2013-46      4
69580    11975  2013-47      3
69581    11975  2013-48      4
69582    11975  2013-49      4
69583    11975  2013-50      6
69587    11975  2014-01      5
69588    11975  2014-02      5
69589    11975  2014-03      3
69590    11975  2014-04      6
69591    11975  2014-05      6
69592    11975  2014-06      6
69593    11975  2014-07      5
69594    11975  2014-08      3
69595    11975  2014-09      4
69596    11975  2014-10      5
69597    11975  2014-11      4
69598    11975  2014-12      4
69599    11975  2014-13      5
69600    11975  2014-14      4
69601    11975  2014-15      6
69602    11975  2014-16      5
69603    11975  2014-17      7
69604    11975  2014-18      7
69605    11975  2014-19      7
69606    11975  2014-20      4
69624    11988  2014-11      4
69626    11988  2014-13      3
69628    11988  2014-15      3
69630    11988  2014-17      3
69633    11988  2014-20      6

[33804 rows x 3 columns]

In [14]:
users['adopted'] = (users['object_id'].isin(adopted))
users.shape

(12000, 11)

In [15]:
print('Number of Adopted Users: {}'.format(len(adopted)))

Number of Adopted Users: 1432


Observation:

* Out of 207917 time stamps, there are 1432 users who logged in 3 or more instances

In [16]:
users.creation_source.value_counts()

ORG_INVITE            4254
GUEST_INVITE          2163
PERSONAL_PROJECTS     2111
SIGNUP                2087
SIGNUP_GOOGLE_AUTH    1385
Name: creation_source, dtype: int64

In [21]:
#users['adopted'] = users['adopted'].replace('True',1)
#users['adopted'] = users['adopted'].replace('False',0)

yes_no_map = {True : 1, "False" : 0}

#users['adopted'] = users['adopted'].map(yes_no_map)
#df["gender"] = df["gender"].map(gender_map)

def map_features(features):
    for feature in features:
        df['adopted'] = df['adopted'].map(yes_no_map)

In [ ]:
users.groupby('creation_source')['adopted'].mean().sort_values(0).plot(kind='barh')
plt.xlabel('Adoption Rate')
plt.ylabel('')
plt.show()

In [22]:
users

object_id        creation_time                 name  \
0              1  2014-04-22 03:53:30       Clausen August   
1              2  2013-11-15 03:45:04        Poole Matthew   
2              3  2013-03-19 23:14:52    Bottrill Mitchell   
3              4  2013-05-21 08:09:28      Clausen Nicklas   
4              5  2013-01-17 10:14:20            Raw Grace   
5              6  2013-12-17 03:37:06        Cunha Eduardo   
6              7  2012-12-16 13:24:32         Sewell Tyler   
7              8  2013-07-31 05:34:02    Hamilton Danielle   
8              9  2013-11-05 04:04:24           Amsel Paul   
9             10  2013-01-16 22:08:03         Santos Carla   
10            11  2013-12-26 03:55:54       Paulsen Malthe   
11            12  2014-04-17 23:48:38      Mathiesen Lærke   
12            13  2014-03-30 16:19:38        Fry Alexander   
13            14  2012-10-11 16:14:33          Rivera Bret   
14            15  2013-07-16 21:33:54          Theiss Ralf   
15            16  2013-02-11 10:09:50           Engel René   
16            17  2014-04-09 14:39:38     Reynolds Anthony   
17            18  2013-08-24 00:26:46     Gregersen Celina   
18            19  2013-05-24 14:56:36       Collins Arlene   
19            20  2014-03-06 11:46:38        Helms Mikayla   
20            21  2013-01-22 12:27:42         Hunter Alfie   
21            22  2014-02-10 06:00:46         Myers Jordan   
22            23  2012-08-17 08:30:27          Souza Alice   
23            24  2013-09-09 22:20:03       Woodruff James   
24            25  2014-02-25 00:11:13          Stone Elise   
25            26  2013-07-28 00:53:32       Wexler Florian   
26            27  2014-01-14 17:35:11          Santos Luís   
27            28  2013-02-11 20:00:25    Weissmuller Marko   
28            29  2013-09-16 02:08:41     Fernandes Evelyn   
29            30  2013-06-04 00:44:25           Egger Leah   
...          ...                  ...                  ...   
11970      11971  2013-05-04 06:54:23    Ferreira Beatrice   
11971      11972  2012-09-13 23:23:11      Penington Jamie   
11972      11973  2012-10-30 03:09:47       Simmons Thomas   
11973      11974  2012-12-26 16:42:08          Scott James   
11974      11975  2013-03-23 11:10:11       Daecher Jürgen   
11975      11976  2013-12-25 22:01:41            Kohl Leah   
11976      11977  2013-09-17 18:58:51         Crary Joshua   
11977      11978  2014-04-23 16:28:06         Castro Pedro   
11978      11979  2013-01-28 09:26:33         Hurst Harley   
11979      11980  2014-02-02 15:23:18  Gloeckner Franziska   
11980      11981  2013-03-05 01:53:48            Fry Tyler   
11981      11982  2013-12-06 14:39:15        Barajas Maria   
11982      11983  2013-07-30 00:23:36           Engel Mike   
11983      11984  2014-02-11 21:26:11           Reed Cerys   
11984      11985  2013-07-08 17:23:26     Jespersen Marcus   
11985      11986  2012-06-24 01:02:14        Ribeiro Clara   
11986      11987  2013-03-15 03:21:01           Reed Poppy   
11987      11988  2014-03-15 11:04:47          Minick John   
11988      11989  2013-09-02 11:24:33       Pereira Emilly   
11989      11990  2013-07-05 21:00:48            Juhl Aase   
11990      11991  2014-04-20 14:22:45        Vaughn Arthur   
11991      11992  2014-04-09 19:39:56     Hermansen Amanda   
11992      11993  2013-03-28 23:24:21      Townsend Isabel   
11993      11994  2012-06-05 04:46:16       Grimstone Joel   
11994      11995  2013-10-21 13:33:17     Werfel Sebastian   
11995      11996  2013-09-06 06:14:15         Meier Sophia   
11996      11997  2013-01-10 18:28:37        Fisher Amelie   
11997      11998  2014-04-27 12:45:16          Haynes Jake   
11998      11999  2012-05-31 11:55:59         Faber Annett   
11999      12000  2014-01-26 08:57:12           Lima Thaís   

                                     email     creation_source  \
0                 AugustCClausen@yahoo.com        GUEST_INVITE   
1                   MatthewPoo

In [ ]:
#1 Update adopted user values

for index, val in adopted1.iteritems():
    print(index,val)
    #users.loc[users['object_id'] == index, 'adopted_user'] = 1 #index is used to connect with users df 'object_id' column
    
#Validate Data
#users.loc[users['object_id'] == 3623]



In [ ]:
#users.creation_source.value_counts()

In [ ]:
#users = users.fillna(0)
users

In [ ]:
# 2. drop irrelevant columns, leaving only "object_id", "opted_in_to_mailing_list", "enabled_for_marketing_drip" and 'adopted_user'

users.drop(['creation_time', 'name', 'email', 'creation_source', 'last_session_creation_time', 'org_id', 'invited_by_user_id'], axis=1, inplace=True)
users

In [ ]:
# 3. Split test and Train data

train = users[:9600] #Training Set
test = users[9600:] #Test Set (based off of Test set being 20% of whole dataset)

df_dummies = pd.get_dummies(users)
X = df_dummies.drop(columns = ['adopted_user'])


In [ ]:
#Drop Adopted user columns from Train and test

trainx = train.drop('adopted_user', axis = 1)

trainy = train['adopted_user']

In [ ]:
#Setting up Testing Dataset

testx = test.drop('adopted_user', axis = 1)

testy = test['adopted_user']

In [ ]:
# 4. Feed these into .fit to build models

##.fit to build Logistic Regression (L.R.) Model

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report
from sklearn.metrics import roc_auc_score,roc_curve,scorer

logit = LogisticRegression(class_weight = 'balanced').fit(trainx, trainy)

logit.fit(trainx,trainy)
predict = logit.predict(testx)
Proba  = logit.predict_proba(testx)

print("Algorithm :",  logit)
print ("\n Classification report : \n",classification_report(testy,predict))
print ("Accuracy   Score : ",accuracy_score(testy,predict))
model_roc_auc = roc_auc_score(testy,predict) 
print ("Area under curve : ",model_roc_auc,"\n")
fpr,tpr,thresholds = roc_curve(testy,Proba[:,1])

Random Forest Classifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics

clf = RandomForestClassifier(random_state=0).fit(trainx, trainy)
clf.fit(trainx,trainy)
y_pred=clf.predict(testx)

print("Algorithm :",  clf)
print ("\n Classification report : \n",classification_report(testy,y_pred))
print ("Accuracy   Score : ",accuracy_score(testy,y_pred))
model_roc_auc = roc_auc_score(testy,y_pred) 
print ("Area under curve : ",model_roc_auc,"\n")
fpr,tpr,thresholds = roc_curve(testy,Proba[:,1])
print("Accuracy:",metrics.accuracy_score(testy, y_pred))

SVM

In [ ]:
from sklearn import svm
model = svm.SVC(kernel='linear')
model.fit(trainx,trainy)
predict = model.predict(testx)
accuracy_score = metrics.accuracy_score(testy, predict)
print("Algorithm :",  model)
print("accuracy score", accuracy_score)


Conclusions from Summary Tables:
Precision identifies the frequency with which a model was correct when predicting the positive class Recall answers the following question: Out of all the possible positive labels, how many did the model correctly identify? F1-score is the weighted average of the precision and recall, where an F1 score reaches its best value at 1 and worst score at 0.
For Classification model metrics/assessment, values closest to 1 were most desirable signaling correct positive predictions. Model that had mostly 1's was Decision Tree model.
Therefore, to re-assess this data, I would choose the Decision Tree model due to it having a high success rate in predicting which users would be classified as 'adopted user'.
Considerations/Recommendations/Further Research
Consideration for code is:
Implement check for 7-days or more period between users' first and third visits in order to truly define user as 'adopted'
Considerations for data:
Have more features to select from in order to better determine/predict an "adopted user". For example, instead of just log-in's, have usage data on certain services they use in order to determine level of user engagement. Otherwise, with current features, there is a lot of extrapolation with how a user gets "adopted" because data on how a user's account gets started does not have a strong relationship with how a user commits to a service provider.